src/GridFS/StreamWrapper.php

```diff
@@ -0,0 +1,182 @@
+        $parsed_path = parse_url($path);
+        $this->databaseName = $parsed_path["host"];
+        $this->identifier = substr($parsed_path["path"], 1);
+        $this->bucketName="fs";
+    }
+
+    public function stream_write($data)
+    {
+        echo strlen($data). "\n";
+        \MongoDB\insert_chunks(NULL, $this->buffer, $data, $this->file["_id"],$this->chunkSizeBytes, $this->filesCollection, $this->chunksCollection, $this->n);
+        return strlen($data);
+    }
+
+    public function stream_read($count) {
+        $out ="";
+        if ($this->dirtyCache) {
+            $out = fread($this->buffer, $count);
+            if(strlen($out) == $count) {
+                return $out;
+            } else {
+                fclose($out);
+                $this->dirtyCache = false;
+            }
+              $this->n++;
+        }
+
+        if($this->file->length <= $this->n){
+            return false;
+        }
+
+        while(strlen($out) < $count && $this ->n <$this->file->length) {
+            $bytes_left = $count - strlen($out);
+            $next = $this->chunksCollection->findOne(["files_id" => $this->file->_id, "n" => $this->n]);
+            $out .= substr($next->data, 0, $bytes_left);
+            $this->n++;
+        }
+        if($bytes_left < strlen($next->data)) {
+            $this->buffer = tmpfile();
+            fwrite($this->buffer, substr($next->data, $bytes_left));
+            $this->dirtyCache =true;
+        }
+        return $out;
+    }
+
+    public function stream_eof(){
+        return $this->n >= $this->file->length && ! $this->dirtyCache;
+    }
+
+    public function stream_close() {
+        if ($this->mode == "w") {
+            rewind($this->buffer);
+            $cached = fread($this->buffer, $this->chunkSizeBytes);
+            fclose($this->buffer);
+            if(strlen($cached) > 0) {
+                $toUpload = ["files_id" => $this->file["_id"], "n" => $this->n, "data" => $cached];
+                $this->chunksCollection->insertOne($toUpload);
+            }
+        }
+        $this->file = array_merge($this->file, ["length" => $this->n+1]);
+        $this->filesCollection->insertOne($this->file);
+    }
+
+    public function stream_open($path, $mode, $options, &$openedPath)
+    {
+        $this->initProtocol($path);
+        $default = stream_context_get_options(stream_context_get_default());
+        $this->filesCollection = new Collection(
+            $default['gridfs']['manager'],
+            sprintf('%s.%s.files', $this->databaseName, $this->bucketName)
+        );
+
+        $this->chunksCollection = new Collection(
+            $default['gridfs']['manager'],
+            sprintf('%s.%s.chunks', $this->databaseName, $this->bucketName)
+        );
+        $this->n = 0;
+        $this->mode = $mode;
+        switch ($this->mode) {
+            case 'w' : return $this ->openWriteStream();
+            case 'r' : return $this ->openReadStream();
+            default: return false;
```

Comment:
AFAIK, the streams are readable or writable, but not both. Additionally, we don't support appending to files (yet), so 'w' and 'r' alone should be sufficient (ignoring 'x' or 'c for now).